flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/FlinkKafkaInternalProducerITCase.java

```diff
@@ -88,20 +92,45 @@ void testInitTransactionId() {
                     reuse.abortTransaction();
                 }
                 assertNumTransactions(i);
-                assertThat(readRecords(TEST_TOPIC).count(), equalTo(i / 2));
+                assertThat(readRecords(TEST_TOPIC).count()).isEqualTo(i / 2);
             }
         }
     }
 
+    @ParameterizedTest
+    @MethodSource("provideTransactionsFinalizer")
+    void testResetInnerTransactionIfFinalizingTransactionFailed(
+            Consumer<FlinkKafkaInternalProducer<?, ?>> transactionFinalizer) {
+        try (FlinkKafkaInternalProducer<String, String> reuse =
+                new FlinkKafkaInternalProducer<>(getProperties(), "dummy")) {
+            reuse.initTransactions();
+            reuse.beginTransaction();
+            reuse.send(new ProducerRecord<>(TEST_TOPIC, "test-value"));
+            reuse.close();
```

Comment:
I'm confused by `close` here. Afterwards, we should do anything with it anymore. Is that your way to ensure that the `commit` fails? How about adding a specific `@VisibleForTesting` method to it instead? In theory after `close` we should have a `checkState(!closed)` on all operations.

Reply:
I updated the test case to start a second producer which fences the first one to fail during committing. PTAL

Reply:
Me gusta.